src/Scriban/Syntax/ScriptVisitor.cs

```diff
@@ -0,0 +1,767 @@
+// Copyright (c) Alexandre Mutel. All rights reserved.
+// Licensed under the BSD-Clause 2 license.
+// See license.txt file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Scriban.Syntax
+{
+    public class ScriptVisitor
+    {
+        private readonly ScriptVisitorContext _context = new ScriptVisitorContext();
+
+        public IScriptVisitorContext Context => _context;
+
+        public ScriptNode Visit(ScriptNode node)
+        {
+            if (node == null) return null;
+
+            using (_context.Push(node))
+            {
+                node = DispatchVisit(node);
+            }
+
+            return node;
+        }
+
+        protected virtual ScriptNode DispatchVisit(ScriptNode node)
```

Comment:
Virtual in case an implementer introduces new types of nodes (which I did in my use case)